astropy/units/format/base.py

```diff
@@ -148,37 +137,49 @@ def to_string(
         ValueError
             If ``fraction`` is not recognized.
         """
+        string_components = cls._to_string_fraction_helper(unit, fraction)
+        if isinstance(string_components, str):
+            return string_components
+        if fraction is True or fraction == "inline":
+            return cls._format_inline_fraction(*string_components)
+        raise ValueError(
+            f"{cls.name!r} format only supports 'inline' fractions, not {fraction=!r}."
+        )
+
+    @classmethod
+    def _to_string_fraction_helper(
+        cls, unit: UnitBase, fraction: bool | Literal["inline", "multiline"]
+    ) -> str | tuple[str, str, str]:
+        # The `unit` is converted to a string if there is no need to decide how
+        # to handle fractions. Otherwise the scale, numerator and denominator
+        # get returned so that `to_string()` could decide how to format them.
+
         # First the scale.  Normally unity, in which case we omit
         # it, but non-unity scale can happen, e.g., in decompositions
         # like u.Ry.decompose(), which gives "2.17987e-18 kg m2 / s2".
-        if unit.scale == 1:
-            s = ""
-        else:
-            s = cls.format_exponential_notation(unit.scale)
-
-        # Now the unit baes, taking care that dimensionless does not have any
-        # (but can have a scale; e.g., u.percent.decompose() gives "0.01").
-        if len(unit.bases):
-            if s:
-                s += cls._scale_unit_separator
-            if fraction:
-                numerator, denominator = utils.get_grouped_by_powers(
-                    unit.bases, unit.powers
-                )
-            else:
-                numerator = list(zip(unit.bases, unit.powers))
-                denominator = []
-            if len(denominator):
-                if len(numerator):
-                    numerator = cls._format_unit_list(numerator)
-                else:
-                    numerator = "1"
-                denominator = cls._format_unit_list(denominator)
-                s = cls._format_fraction(s, numerator, denominator, fraction=fraction)
+        s = "" if unit.scale == 1 else cls.format_exponential_notation(unit.scale)
+
+        # dimensionless does not have any bases, but can have a scale;
+        # e.g., u.percent.decompose() gives "0.01".
+        if not unit.bases:
+            return s
+
+        if s:
+            s += cls._scale_unit_separator
+        if not fraction:
+            return s + cls._format_unit_list(zip(unit.bases, unit.powers, strict=True))
+        positive = []
+        negative = []
+        for base, power in zip(unit.bases, unit.powers, strict=True):
+            if power < 0:
+                negative.append((base, -power))
+            elif power > 0:
+                positive.append((base, power))
             else:
-                s += cls._format_unit_list(numerator)
-
-        return s
+                raise ValueError("Unit with 0 power")
```

Comment:
This is very defensive programming, I'm not convinced it is actually needed.

Reply:
This has been shaping my thinking lately: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/
I agree we shouldn't have "shotgun [validation]" but we should have well-crafted parsing...